dicompylercore/dose.py

```diff
@@ -0,0 +1,372 @@
+from dicompylercore.config import dicompyler_uid_prefix_rtdose, scipy_available
+from datetime import datetime
+from pydicom.sequence import Sequence
+from pydicom.dataset import Dataset
+
+if scipy_available:
+    from scipy.ndimage import map_coordinates
+
+
+class DoseGrid:
+    """
+    Class to easily access commonly used attributes of a DICOM dose grid and
+    make modifications
+
+    Example: Add two dose grids
+        grid_1 = DoseGrid(dose_file_1)
+        grid_2 = DoseGrid(dose_file_2)
+        grid_sum = grid_1 + grid_2
+        grid_sum.save_dcm(some_file_path)
+
+    """
+
+    def __init__(self, rt_dose, order=1):
+        """ Initialization of a DoseGrid from a DICOM-RT Dose file or dataset.
+
+        Parameters
+        ----------
+        rt_dose : pydicom Dataset or filename
+            DICOM RT Dose used to determine the structure dose grid data.
+        order : int, optional
+            The order of the spline interpolation.
+            0: the nearest grid point, 1: trilinear, 2 to 5: spline
+        """
+
+        self.ds = dicomparser.DicomParser(rt_dose).ds
+        self.order = order
+
+        self.summation_type = None
+        self.sop_class_uid = self.ds.SOPClassUID
+        self.sop_instance_uid = self.ds.SOPInstanceUID
+        self.other_sop_class_uid = None
+        self.other_sop_instance_uid = None
+
+        # TODO: Raise error if not rtdose?
+        if self.ds.Modality == "RTDOSE":
+            self.x_axis = (
+                np.arange(self.ds.Columns) * self.ds.PixelSpacing[0]
+                + self.ds.ImagePositionPatient[0]
+            )
+            self.y_axis = (
+                np.arange(self.ds.Rows) * self.ds.PixelSpacing[1]
+                + self.ds.ImagePositionPatient[1]
+            )
+            self.z_axis = (
+                np.array(self.ds.GridFrameOffsetVector)
+                + self.ds.ImagePositionPatient[2]
+            )
+
+            # x and z are swapped in the pixel_array
+            pixel_array = self.ds.pixel_array * self.ds.DoseGridScaling
+            self.dose_grid = np.swapaxes(pixel_array, 0, 2)
+
+    ####################################################
+    # Basic properties
+    ####################################################
+    @property
+    def shape(self):
+        """Get the x, y, z dimensions of the dose grid"""
+        return tuple(
+            [self.ds.Columns, self.ds.Rows, len(self.ds.GridFrameOffsetVector)]
+        )
+
+    @property
+    def axes(self):
+        """Get the x, y, z axes of the dose grid (in mm)"""
+        return [self.x_axis, self.y_axis, self.z_axis]
+
+    @property
+    def scale(self):
+        """Get the dose grid resolution (xyz)"""
+        return np.array(
+            [
+                self.ds.PixelSpacing[0],
+                self.ds.PixelSpacing[1],
+                self.ds.GridFrameOffsetVector[1]
+                - self.ds.GridFrameOffsetVector[0],
+            ]
```

Comment:
Good thought.  The `scale` property currently is only used invoked with `DoseGrid.interp_entire_grid`. The new `NotImplementedError` raised in `scale` will catch this. I'm not sure what if `scale` is useful otherwise?

Or perhaps raising a `UserWarning` in `DoseGrid.scale` is more appropriate?  Then raise the `NotImplementedError` in `DoseGrid.interp_entire_grid`?

https://github.com/cutright/dicompyler-core/blob/a5bd38a4458411744f1398e1a644eb4c32d035c3/dicompylercore/dose.py#L102-L117